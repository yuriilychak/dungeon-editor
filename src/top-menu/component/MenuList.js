import {withStyles} from "@material-ui/core";
import MenuList from "@material-ui/core/MenuList";

const styles = theme => ({
    root: {
        backgroundColor: theme.palette.primary.main,
        borderRadius: 0,
        padding: 0
    }
});

export default withStyles(styles)(MenuList);
