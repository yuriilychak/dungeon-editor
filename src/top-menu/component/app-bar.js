import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';

const styles = theme => ({
    root: {
        backgroundColor: theme.palette.primary.dark
    }
});

export default withStyles(styles)(AppBar);
