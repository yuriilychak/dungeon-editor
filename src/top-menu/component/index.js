export { default as TopMenu } from "./top-menu";
