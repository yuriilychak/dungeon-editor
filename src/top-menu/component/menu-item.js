import { withStyles } from "@material-ui/core";
import { default as Element } from "@material-ui/core/MenuItem";
import React, {memo} from "react";
import { useTranslation } from 'react-i18next';
import Check from "@material-ui/icons/Check";

const height = 18;
const minWidth = 128;
const horizontalPadding = 8;

const menuItemStyles = theme => ({
    root: {
        backgroundColor: theme.palette.primary.main,
        borderRadius: 0,
        padding: `2px ${horizontalPadding}px`,
        height: height,
        minHeight: height,
        minWidth: minWidth,
        color: theme.palette.primary.contrastText,
        fontSize: "0.7rem",
        fontFamily: theme.typography.fontFamily,
        display: "grid",
        gridTemplateColumns: "1fr 32px"
    }
});

const iconStyle = {
    root: {
        marginRight: horizontalPadding,
        fontSize: "0.7rem"
    }
};

const leftRowStyle = {
    root: {
        float: "left",
        textAlign: "left"
    }
};

const rightRowStyle = {
    root: {
        float: "right",
        textAlign: "right",
        color: "#cccccc",
    }
};

const Icon = withStyles(iconStyle)(Check);
const Item = withStyles(menuItemStyles)(Element);
const createRow = style => withStyles(style)(props => <div className={props.classes.root}>{props.children}</div>);
const LeftRow = createRow(leftRowStyle);
const RightRow = createRow(rightRowStyle);
const EmptyIcon = () => (<span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span>);

export const MenuItem = props => {
    const { t } = useTranslation();
    const { onClick, id, locale, hotKey, isSelected} = props;

    return (
        <Item onClick={() => { onClick(id) }}>
            <LeftRow>
                {isSelected ? <Icon/> : <EmptyIcon/>}
                {t(locale)}
            </LeftRow>
            <RightRow>{hotKey}</RightRow>
        </Item>
    )
};


export default memo(MenuItem);
