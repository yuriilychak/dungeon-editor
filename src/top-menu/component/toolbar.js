import { withStyles } from '@material-ui/core/styles';
import Toolbar from "@material-ui/core/Toolbar";
import { ITEM_SIZE, INDENT_SIZE } from "../../constant/ui";

export default withStyles({
    root: {
        paddingLeft: INDENT_SIZE.L,
        paddingRight: INDENT_SIZE.L,
        height: ITEM_SIZE.SMALL,
        minHeight: ITEM_SIZE.SMALL,
        zIndex: 1,
        "@media (min-width: 600px)": {
            minHeight: ITEM_SIZE.SMALL
        }
    }
})(Toolbar);
