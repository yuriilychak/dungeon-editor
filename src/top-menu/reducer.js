import TOP_MENU from "./state";
import SUBSECTIONS from "./enum/subsection";

import StaticData from "./data/index.json";
import MENU_STATE from "./enum/menu-state";
import { handleAction } from "../helpers";

export const initialState = {
    staticData: StaticData,
    openMenu: MENU_STATE.NONE,
    toggledSections: [
        SUBSECTIONS.ANCHOR,
        SUBSECTIONS.GUIDES,
        SUBSECTIONS.RULER,
        SUBSECTIONS.CHANGE_SIZE
    ]
};

const { tabs } = StaticData;
const toggleIds = [];

tabs.forEach( tab =>
    tab.sections.forEach(
        section => {
            if (section.isToggle) {
                toggleIds.push(section.id);
            }
        }
    )
);

const actionHandlers = {
    [TOP_MENU.CHANGE]: (state, openMenu) => ({ ...state, openMenu }),
    [TOP_MENU.SECTION_TOGGLE]: (state, sectionId) => {
        if (!toggleIds.includes(sectionId)) {
            return state;
        }

        const toggledSections = [...state.toggledSections];

        let index;

        if (sectionId === SUBSECTIONS.CHANGE_SIZE || sectionId === SUBSECTIONS.CHANGE_SCALE) {

            if (toggledSections.includes(sectionId)) {
                return state;
            }

            const removeId = sectionId === SUBSECTIONS.CHANGE_SIZE ?
                SUBSECTIONS.CHANGE_SCALE : SUBSECTIONS.CHANGE_SIZE;
            index = toggledSections.indexOf(removeId);

            if (index !== -1) {
                toggledSections.splice(index, 1);
            }

            toggledSections.push(sectionId);
        }
        else {
            index = toggledSections.indexOf(sectionId);

            if (index === -1) {
                toggledSections.push(sectionId);
            }
            else {
                toggledSections.splice(index, 1);
            }
        }

        return {
            ...state,
            toggledSections
        };
    }
};

export default function(state = initialState, action) {
    return handleAction(state, actionHandlers, action);
}
