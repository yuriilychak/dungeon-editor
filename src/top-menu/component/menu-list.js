import {withStyles} from "@material-ui/core";
import MenuList from "@material-ui/core/MenuList";

const styles = theme => ({
    root: {
        backgroundColor: theme.palette.primary.main,
        borderRadius: 0,
        padding: 0,
        paddingBottom: 8
    }
});

export default withStyles(styles)(MenuList);
