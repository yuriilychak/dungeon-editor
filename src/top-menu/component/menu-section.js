import React, { memo, useState } from 'react';
import { useTranslation } from "react-i18next";

import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Grow from '@material-ui/core/Grow';
import Popper from '@material-ui/core/Popper';

import Button from "./button";
import MenuItem from './menu-item';
import MenuList from './menu-list';
import MenuBackground from "./menu-background";

const MenuSection = props => {
    const { t } = useTranslation();
    const {locale, onOpen, onClose, onSelectSection, isOpen, sections, id, toggledSections} = props;
    const [anchorEl, setAnchorEl] = useState(null);

    function handleClick(event) {
        setAnchorEl(event.currentTarget);
        onOpen(id);
    }

    function handleClose() {
        setAnchorEl(null);
        onClose();
    }

    const sectionComponents = sections.map(section =>
        (<MenuItem
            onClick={ onSelectSection }
            key={section.id}
            isSelected={section.isToggle && toggledSections && toggledSections.includes(section.id)}
            {...section}
        />)
    );
    const growId = "menu-list-grow";

    return (
        <div>
            <Button
                aria-owns={isOpen ? growId : undefined}
                aria-haspopup="true"
                onClick={handleClick}
            >
               {t(locale)}
            </Button>
            <Popper
                placement="bottom-start"
                open={isOpen}
                anchorEl={anchorEl}
                transition
                disablePortal
            >
                {({ TransitionProps }) => (
                    <Grow
                        {...TransitionProps}
                        id={growId}
                        style={{ transformOrigin: 'left bottom' }}
                    >
                        <MenuBackground>
                            <ClickAwayListener onClickAway={handleClose}>
                                <MenuList>
                                    {sectionComponents}
                                </MenuList>
                            </ClickAwayListener>
                        </MenuBackground>
                    </Grow>
                )}
            </Popper>
        </div>
    );
};


export default memo(MenuSection);
