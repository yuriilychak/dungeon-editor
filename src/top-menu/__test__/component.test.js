import AppBar from "../component/app-bar";
import Button from "../component/button";
import MenuBackground from "../component/menu-background";
import { MenuItem } from "../component/menu-item";
import MenuList from "../component/menu-list";
import { MenuSection } from "../component/menu-section";
import Toolbar from "../component/toolbar";
import TopMenu from "../component/top-menu";
import React from "react";
import reducer from "../reducer";
import { createMount, createShallow } from "@material-ui/core/test-utils";
import {getInitialState} from "../../../test_templates";
import {UI_SECTION} from "../../enum";

document.createRange = jest.fn();

jest.mock("@material-ui/core/Popper", () => ({children}) => (
    <div>
        {children}
    </div>
));

describe("top-menu test",()=> {
    const initialState = getInitialState(reducer, UI_SECTION.TOP_MENU);
    it ( "top-menu snapshot", () => {
        const wrapper = createShallow()(
            <TopMenu {...initialState} />
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    it ( "Button snapshot", () => {
        const wrapper = createShallow()(
            <Button>test</Button>
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    it ( "AppBar snapshot", () => {
        const wrapper = createShallow()(
            <AppBar>
                <Button>test1</Button>
                <Button>test2</Button>
                <Button>test3</Button>
            </AppBar>
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    it ( "ToolBar snapshot", () => {
        const wrapper = createShallow()(
            <Toolbar>
                <Button>test1</Button>
                <Button>test2</Button>
                <Button>test3</Button>
            </Toolbar>
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    it ( "MenuBackground snapshot", () => {
        const wrapper = createShallow()(
            <MenuBackground>
                <Button>test1</Button>
                <Button>test2</Button>
                <Button>test3</Button>
            </MenuBackground>
        );

        expect(wrapper.html()).toMatchSnapshot();
    });
    it ( "MenuList snapshot", () => {
        const wrapper = createShallow()(
            <MenuList>
                <Button>test1</Button>
                <Button>test2</Button>
                <Button>test3</Button>
            </MenuList>
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    describe("MenuItem test",() => {
        const props = {
            onClick: jest.fn(),
            id: 0,
            locale: "test",
            hotKey: "Ctr+Shift",
            isSelected: false
        };

        const wrapper = createMount()(
            <MenuItem
                {...props}
            />
        );

        it ( "snapshot default", () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        it ( "snapshot selected", () => {
            const selectedProps = {
                ...props,
                isSelected: true
            };
            const selectedWrapper = createShallow()(
                <MenuItem
                    {...selectedProps}
                />
            );
            expect(selectedWrapper.html()).toMatchSnapshot();
        });

        it ( "click", () => {
            jest.spyOn(wrapper.props(), "onClick");

            wrapper.find("li").first().simulate("click");

            const onClick = wrapper.props().onClick;

            expect(onClick).toHaveBeenCalled();
            expect(onClick).toHaveBeenCalledWith(0);
        });
    });

    describe("MenuSection test", () => {
        const onOpen = jest.fn();
        const props = {
            "id": 1,
            "locale": "TopMenu_File",
            "useToggles": false,
            "sections": [
                {
                    "id": 0,
                    "locale": "TopMenu_NewProject",
                    "hotKey": "Ctr+N",
                    "isToggle": false
                },
                {
                    "id": 1,
                    "locale": "TopMenu_OpenProject",
                    "hotKey": "Ctr+O",
                    "isToggle": false
                },
                {
                    "id": 2,
                    "locale": "TopMenu_SaveAll",
                    "hotKey": "Ctr+S",
                    "isToggle": false
                },
                {
                    "id": 3,
                    "locale": "TopMenu_Export",
                    "hotKey": "Ctr+E",
                    "isToggle": true
                },
                {
                    "id": 17,
                    "locale": "TopMenu_Publish",
                    "hotKey": "Ctr+P",
                    "isToggle": false
                },
                {
                    "id": 4,
                    "locale": "TopMenu_Quit",
                    "hotKey": "Ctr+Q",
                    "isToggle": false
                }
            ],
            onOpen,
            onClose: jest.fn(),
            toggledSections: [3],
            isOpen: true
        };

        const wrapper = createMount()(
            <MenuSection
                {...props}
            />
        );

        it ("snapshot open", () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        it ("snapshot closed", () => {
            const closedProps = {
                ...props,
                isOpen: false
            };
            const closedWrapper = createShallow()(
                <MenuSection
                    {...closedProps}
                />
            );
            expect(closedWrapper.html()).toMatchSnapshot();
        });

        it ("click", () => {
            wrapper.find("button").at(0).simulate("click");

            expect(onOpen).toHaveBeenCalled();
            expect(onOpen).toHaveBeenCalledWith(1);
        });
    });
});
