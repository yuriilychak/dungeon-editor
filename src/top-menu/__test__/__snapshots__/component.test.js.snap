// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`top-menu test AppBar snapshot 1`] = `"<header class=\\"MuiPaper-root MuiAppBar-root WithStyles(ForwardRef(AppBar))-root-6 MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed MuiPaper-elevation4\\"><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test1</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test2</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test3</span></button></header>"`;

exports[`top-menu test Button snapshot 1`] = `"<button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test</span></button>"`;

exports[`top-menu test MenuBackground snapshot 1`] = `"<div class=\\"MuiPaper-root WithStyles(ForwardRef(Paper))-root-8 MuiPaper-elevation1 MuiPaper-rounded\\"><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test1</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test2</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test3</span></button></div>"`;

exports[`top-menu test MenuItem test snapshot default 1`] = `"<button><div class=\\"Component-root-2\\"><span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span>test</div><div class=\\"Component-root-3\\">Ctr+Shift</div></button>"`;

exports[`top-menu test MenuItem test snapshot selected 1`] = `"<button><div class=\\"Component-root-2\\"><svg class=\\"MuiSvgIcon-root pure(CheckIcon)-root-10\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z\\"></path></svg>test</div><div class=\\"Component-root-3\\">Ctr+Shift</div></button>"`;

exports[`top-menu test MenuList snapshot 1`] = `"<ul class=\\"MuiList-root ForwardRef(MenuList)-root-9 MuiList-padding\\" role=\\"menu\\" tabindex=\\"-1\\"><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test1</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test2</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test3</span></button></ul>"`;

exports[`top-menu test MenuSection test snapshot closed 1`] = `"<div><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">File</span></button><div></div></div>"`;

exports[`top-menu test MenuSection test snapshot open 1`] = `"<div><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">File</span></button><div></div></div>"`;

exports[`top-menu test ToolBar snapshot 1`] = `"<div class=\\"MuiToolbar-root WithStyles(ForwardRef(Toolbar))-root-7 MuiToolbar-regular MuiToolbar-gutters\\"><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test1</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test2</span></button><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">test3</span></button></div>"`;

exports[`top-menu test top-menu snapshot 1`] = `"<header class=\\"MuiPaper-root MuiAppBar-root WithStyles(ForwardRef(AppBar))-root-6 MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\"><div class=\\"MuiToolbar-root WithStyles(ForwardRef(Toolbar))-root-7 MuiToolbar-regular MuiToolbar-gutters\\"><div><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">File</span></button><div></div></div><div><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">Edit</span></button><div></div></div><div><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">View</span></button><div></div></div><div><button><span class=\\"MuiButton-label WithStyles(ForwardRef(Button))-label-5\\">Help</span></button><div></div></div></div></header>"`;
